### PR TITLE
Add from_base64url Presto Function

### PR DESCRIPTION
**Pull request by ashokku2022**
Resolves: https://github.com/facebookincubator/velox/issues/4336

### CODE REVIEW

File: velox/common/encode/Base64.cpp [hunk 1]
```diff
@@ -428,6 +438,10 @@ std::string Base64::encode_url(const folly::IOBuf* data) {
   return encodeImpl(IOBufWrapper(data), kBase64UrlCharset, false);
 }
 
+static void decode(const char* data, size_t size, char* output) {
```
Comment (aditi-pandit) [inline velox/common/encode/Base64.cpp:441]: This function doesn't seem to have a body. Is this something left over ? Please delete it if not needed.

File: velox/common/encode/Base64.cpp [hunk 2]
```diff
@@ -305,6 +305,16 @@ void Base64::decode(const char* data, size_t size, char* output) {
   Base64::decode(data, size, output, out_len);
 }
 
+void Base64::decode_url(const char* data, size_t size, char* output) {
```
Comment (aditi-pandit) [inline velox/common/encode/Base64.cpp:308]: There are few decode_url functions below around line 445. Can these functions be consolidated with those ?

File: velox/functions/lib/string/StringImpl.h [hunk 3]
```diff
@@ -419,6 +419,21 @@ FOLLY_ALWAYS_INLINE bool fromBase64(
...
 
+template <typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE bool fromBase64Url(
+    TOutString& output,
+    const TInString& input) {
+  try {
+    auto inputSize = input.size();
+    output.resize(
+        encoding::Base64::calculateDecodedSize(input.data(), inputSize));
+    encoding::Base64::decode_url(input.data(), input.size(), output.data());
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/functions/lib/string/StringImpl.h]: addressed

File: velox/common/encode/Base64.cpp (hunk 1 above)
Comment (ashokku2022) [reply velox/common/encode/Base64.cpp]: deleted

File: velox/common/encode/Base64.cpp (hunk 2 above)
Comment (ashokku2022) [reply velox/common/encode/Base64.cpp]: consolidated

**Comment by ashokku2022** [comment]
Thanks @aditi-pandit for your feedback. I have accommodated all the suggestions. 
Not sure why the format check is failing though, any suggestions?

**Comment by majetideepak** [comment]
Can you run `make format-fix` to fix the formatting?

File: velox/common/encode/Base64.cpp [hunk 4]
```diff
@@ -428,6 +428,16 @@ std::string Base64::encode_url(const folly::IOBuf* data) {
   return encodeImpl(IOBufWrapper(data), kBase64UrlCharset, false);
 }
 
+void Base64::decode_url(const char* data, size_t size, char* output) {
+  size_t out_len = size / 4 * 3;
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/common/encode/Base64.cpp]: Have updated code and removed out_len and got it calculated in output.resize().

**Comment by aditi-pandit** [comment]
@ashokku2022 : The linux-build failure shows a failure for fromBase64Url test. Please can you investigate.

173: [  FAILED  ] StringFunctionsTest.fromBase64Url

**Comment by ashokku2022** [comment]
@aditi-pandit I ran fuzzer as follow, here is the output
`velox_expression_fuzzer_test  --seed 123456 --duration_sec 60 --repro_persist_path --logtostderr=1 --minloglevel=0  --only from_base64url  --v=1 2>&1 | tee fuzzertest.out`

from_base64url
```
E20230322 06:05:01.909464 6389639 Exceptions.h:68] Line: /Users/ashokku/velox/./velox/functions/lib/string/StringImpl.h:435, Function:fromBase64Url, Expression:  Base64::decode() - invalid input string: string length cannot be 1 more than a multiple of 4., Source: USER, ErrorCode: INVALID_ARGUMENT
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

Seeing 0, I was bit unclear about its output so I ran fuzzer for already implemented scalar function like `from_base64` and `from_hex` and output for them were similar. Please let me know if I am doing something wrong. Thanks!



from_base64
```
E20230322 06:14:26.242478 6396509 Exceptions.h:68] Line: /Users/ashokku/velox/./velox/functions/lib/string/StringImpl.h:417, Function:fromBase64, Expression:  Base64::decode() - invalid input string: invalid characters, Source: USER, ErrorCode: INVALID_ARGUMENT
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

from_hex
```
E20230322 06:18:54.691723 6400229 Exceptions.h:68] Line: /Users/ashokku/velox/./velox/functions/lib/string/StringImpl.h:371, Function:fromHex, Expression:  Invalid hex character: z, Source: USER, ErrorCode: INVALID_ARGUMENT
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

**Comment by ashokku2022** [comment]
@aditi-pandit looks like fuzzer test failed in linux-build, however, locally it succeed for me. Can you look into it please?

```
I20230323 11:51:06.154752 7636365 ExpressionFuzzer.cpp:1158] ==============================> Done with iteration 17556
I20230323 11:51:06.154791 7636365 ExpressionFuzzer.cpp:972] ==============================> Top 1 by number of rows processed
I20230323 11:51:06.154806 7636365 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230323 11:51:06.154819 7636365 ExpressionFuzzer.cpp:978] from_base64url 17557 100.00% 305767
I20230323 11:51:06.154836 7636365 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I20230323 11:51:06.154850 7636365 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230323 11:51:06.154863 7636365 ExpressionFuzzer.cpp:991] from_base64url 17557 100.00% 305767
I20230323 11:51:06.154878 7636365 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I20230323 11:51:06.154891 7636365 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230323 11:51:06.154904 7636365 ExpressionFuzzer.cpp:1010] from_base64url 17557 100.00% 305767
```

**Comment by aditi-pandit** [comment]
Thanks @ashokku2022 . Please can you squash your commits if possible.

**Pull request closed without merging**